platform/src/syscalls_impl.rs

```diff
@@ -33,6 +35,113 @@ impl<S: RawSyscalls> Syscalls for S {
+            core::mem::forget(exit);
+        }
+
+        // Inner function that does the majority of the work. This is not
+        // monomorphized over DRIVER_NUM and SUBSCRIBE_NUM to keep code size
+        // small.
+        //
+        // Safety: upcall_fcn must be kernel_upcall<S, IDS, U> and upcall_data
+        // must be a reference to an instance of U that will remain valid as
+        // long as the 'scope lifetime is alive. Can only be called if a
+        // Subscribe<'scope, S, driver_num, subscribe_num> exists.
+        unsafe fn inner<S: Syscalls, CONFIG: subscribe::Config>(
+            driver_num: u32,
+            subscribe_num: u32,
+            upcall_fcn: Register,
+            upcall_data: Register,
+        ) -> Result<(), ErrorCode> {
+            // Safety: syscall4's documentation indicates it can be used to call
+            // Subscribe. These arguments follow TRD104. kernel_upcall has the
+            // required signature. This function's preconditions mean that
+            // upcall is a reference to an instance of U that will remain valid
+            // until the 'scope lifetime is alive The existence of the
+            // Subscribe<'scope, Self, DRIVER_NUM, SUBSCRIBE_NUM> guarantees
+            // that if this Subscribe succeeds then the upcall will be cleaned
+            // up before the 'scope lifetime ends, guaranteeing that upcall is
+            // still alive when kernel_upcall is invoked.
+            let [r0, r1, _, _] = unsafe {
+                S::syscall4::<{ syscall_class::SUBSCRIBE }>([
+                    driver_num.into(),
+                    subscribe_num.into(),
+                    upcall_fcn,
+                    upcall_data,
+                ])
+            };
+
+            let return_variant: ReturnVariant = r0.as_u32().into();
+            if return_variant == return_variant::FAILURE_2_U32 {
+                // Safety: TRD 104 guarantees that if r0 is Failure with 2 U32,
+                // then r1 will contain a valid error code. ErrorCode is
+                // designed to be safely transmuted directly from a kernel error
+                // code.
+                Err(unsafe { core::mem::transmute(r1.as_u32() as u16) })
+            } else {
+                // r0 indicates Success with 2 u32s. Confirm the null upcall was
+                // returned, and it if wasn't then call the configured function.
+                // We're relying on the optimizer to remove this branch if
+                // returned_nonnull_upcall is a no-op.
+                let returned_upcall: *const () = r1.into();
+                #[allow(clippy::zero_ptr)]
+                if returned_upcall != 0 as *const () {
+                    CONFIG::returned_nonnull_upcall(driver_num, subscribe_num);
+                }
+                Ok(())
+            }
+        }
+
+        let upcall_fcn = (kernel_upcall::<S, IDS, U> as usize).into();
+        let upcall_data = (upcall as *const U).into();
```

Comment:
I used `*const U` because it is consistent with the `Upcall` trait definition, which passes `&self`. The trait cannot use `&mut self`, because `self` will be aliased (the kernel upcall and the application code that is polling for completion will both access the upcall).

It doesn't really matter though.

Reply:
I agree that is doesn't really matter for soundness, but I do think is illustrates the expected data flow a little better. I lean more towards `*mut`, but could definitely live with `*const`

Reply:
Turns out it does matter:

```
error[E0606]: casting `&'scope U` as `*mut U` is invalid
   --> platform/src/syscalls_impl.rs:138:27
    |
138 |         let upcall_data = (upcall as *mut U).into();
```

I'd prefer `(upcall as *const U).into()` over `(upcall as *const U as *mut U).into()`, so I left it as-is.

Reply:
That's fine. Maybe just add a comment that we expect that this data will get modified? If you don't feel that is necessary though, I am okay with that.

Reply:
I don't think that's necessary.